### PR TITLE
Profiling - support Chakra & Kineto trace dumping

### PR DESCRIPTION
**Pull request by lilyw97**
# What does this PR do ?

**Add support of Chakra trace & Kineto trace dumping by `torch.profiler.profile`,  take nsys profiling as PR reference**

**Collection**: [Note which collection this PR will affect] **All collections which is based on modelPT**

# Changelog 
- Add specific line by line info of high level changes in this PR.
modePT.py:1747  Add setup chakra profiling
modePT.py:1917  Start chakra profiling at batch start callback
modePT.py:1917  End chakra profiling at batch end callback

# Usage
* You can potentially add a usage example below
NeMo-Framework-Launcher
```python
# Add a code snippet demonstrating how to use this 
python3 ${NEMO_FRAMEWORK_LAUNCHER_DIR}/launcher_scripts/main.py \
training=gpt3/5b \
...
training.trainer.num_nodes=2 \
+training.model.optim.grad_sync_dtype=bf16 \
+training.model.chakra_profile.enabled=True \
+training.model.chakra_profile.start_step=0 \
+training.model.chakra_profile.end_step=10 \
+training.model.chakra_profile.warmup_steps=0 \
+training.model.chakra_profile.active_steps=1 \
+training.model.chakra_profile.trace_dir=${NEMO_FRAMEWORK_LAUNCHER_DIR}results/chakra/ \
```

# GitHub Actions CI

The Jenkins CI system has been replaced by GitHub Actions self-hosted runners.

The GitHub Actions CI will run automatically when the "Run CICD" label is added to the PR.
To re-run CI remove and add the label again.
To run CI on an untrusted fork, a NeMo user with write access must first click "Approve and run".

# Before your PR is "Ready for review"
**Pre checks**:
- [x] Make sure you read and followed [Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md)
- [ ] Did you write any new necessary tests?
- [ ] Did you add or update any necessary documentation?
- [ ] Does the PR affect components that are optional to install? (Ex: Numba, Pynini, Apex etc)
  - [ ] Reviewer: Does the PR have correct import guards for all optional libraries?
  
**PR Type**:
- [x] New Feature
- [ ] Bugfix
- [ ] Documentation

If you haven't finished some of the above items you can still open "Draft" PR.

## Who can review?

Anyone in the community is free to review the PR once the checks have passed. 
[Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md) contains specific people who can review PRs to various areas.

# Additional Information
* Related to # (issue)

### CODE REVIEW

File: nemo/core/classes/modelPT.py [hunk 1]
```diff
@@ -1735,6 +1744,75 @@ def update_save_restore_connector(cls, save_restore_connector):
         else:
             setattr(cls, '_save_restore_connector', save_restore_connector)
 
+    def _setup_chakra_profiling(self):
+        """ Enables chakra profiling
+            To use, add the following optoins to the model config:
```
Comment (maanug-nv) [inline nemo/core/classes/modelPT.py:1749]: nit, typo optoins->options

File: nemo/core/classes/modelPT.py [hunk 2]
```diff
@@ -1735,6 +1744,75 @@ def update_save_restore_connector(cls, save_restore_connector):
...
+        """
+        if self.cfg.get('chakra_profile', None) is not None:
+            if self.cfg.chakra_profile.get('enabled', False):
+                self._chakra_profile_enabled = True
+                self._chakra_profile_start_step = self.cfg.chakra_profile.get('start_step', 0)
+                self._chakra_profile_end_step = self.cfg.chakra_profile.get('end_step', 0)
+
+                if self.cfg.chakra_profile.get('trace_dir', None) is None:
+                    raise ValueError(
+                        f'chakra profile output path ({self._memory_profile_output_path}) is not set.'
```
Comment (maanug-nv) [inline nemo/core/classes/modelPT.py:1767]: Seems this hasn't been assigned yet? Also can we do '_chakra_output_path' for consistency, since 'memory_profile' prefix is used for all the Nsys attributes.

File: nemo/core/classes/modelPT.py [hunk 3]
```diff
@@ -1841,9 +1919,16 @@ def on_train_batch_start(self, batch: Any, batch_idx: int, unused: int = 0) -> O
         https://pytorch-lightning.readthedocs.io/en/stable/common/lightning_module.html#on-train-batch-start
         We use it here to enable nsys profiling and dynamic freezing.
         """
-
         # nsys profiling
```
Comment (maanug-nv) [inline nemo/core/classes/modelPT.py:1922]: can we move this comment

File: nemo/core/classes/modelPT.py [hunk 4]
```diff
@@ -1735,6 +1744,75 @@ def update_save_restore_connector(cls, save_restore_connector):
...
+                else:
+                    raise ValueError(f'chakra end_step must be of type int. Found: {type(self._chakra_profile_end_step)}')
+
+                if self._chakra_profile_end_step >= self._chakra_profile_start_step:
+                    pass
+                else:
+                    raise ValueError(f'chakra end_step must be greater than or equal to chakra start_step')
+
+                if self.cfg.nsys_profile.get('enabled', False):
+                    logging.warning(f'Profiling with Nsys enabled will result in the Kineto trace losing GPU performance data.')
```
[the remark this answers is not in the document]
Comment (maanug-nv) [reply nemo/core/classes/modelPT.py:1801]: @ericharper if you have a different opinion on this, let us know

File: nemo/core/classes/modelPT.py [hunk 5]
```diff
@@ -1735,6 +1744,75 @@ def update_save_restore_connector(cls, save_restore_connector):
...
+                self._prof = torch.profiler.profile(
+                    activities=[torch.profiler.ProfilerActivity.CPU,torch.profiler.ProfilerActivity.CUDA,],
+                    schedule=torch.profiler.schedule(
+                        wait=0,
+                        warmup=warmup_steps,
+                        active=active_steps
+                    ),
+                    record_shapes=True,
+                    execution_trace_observer=self._et,
+                    experimental_config=_ExperimentalConfig(enable_cuda_sync_events=True)
```
Comment (sanshang-nv) [inline nemo/core/classes/modelPT.py:1813]: Why we need `record_shapes` and `experimental_config` here? Are them necessary?

File: nemo/core/classes/modelPT.py (hunk 2 above)
Comment (lilyw97) [reply nemo/core/classes/modelPT.py]: Resolved in the latest commit, thank you for the reminder

File: nemo/core/classes/modelPT.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (lilyw97) [reply nemo/core/classes/modelPT.py:1802]: Seems even with Nsys enabled, torch could still dump the CPU information in Kineto trace, so it seems not like an absolute opposite option, Hi @sanshang-nv , how do you think so?

File: nemo/core/classes/modelPT.py (hunk 5 above)
Comment (lilyw97) [reply nemo/core/classes/modelPT.py]: I removed these 2 options in the latest commit

File: nemo/core/classes/modelPT.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (lilyw97) [reply nemo/core/classes/modelPT.py:1802]: conclusion: raise exception when nsys is also enabled.

**Comment by lilyw97** [comment]
Hi @maanug-nv , I have fixed all the comments, can you help to review it again? I did `python setup.py style --fix` locally, and it returns successfully, but the reformat job seems failed, looks like it's a flow issue, I think it looks ready to be merged in.

File: nemo/core/classes/modelPT.py [hunk 6]
```diff
@@ -1839,11 +1916,17 @@ def on_train_start(self):
...
         https://pytorch-lightning.readthedocs.io/en/stable/common/lightning_module.html#on-train-batch-start
-        We use it here to enable nsys profiling and dynamic freezing.
+        We use it here to enable profiling and dynamic freezing.
         """
-
-        # nsys profiling
         if self.device.type == 'cuda':
+            if hasattr(self, '_chakra_profile_enabled'):
+                if self._chakra_profile_enabled and not self._chakra_profile_in_progress:
+                    if self.trainer.global_step >= self._chakra_profile_start_step and self.trainer.global_step < self._chakra_profile_end_step:
```
Comment (maanug-nv) [inline nemo/core/classes/modelPT.py:1924]: based on line 1796 (`if self._chakra_profile_end_step >= self._chakra_profile_start_step`), we want to support profiling for just a single step, right? so we should do `self.trainer.global_step <= self._chakra_profile_end_step`

Comment (lilyw97) [reply nemo/core/classes/modelPT.py]: Hi @maanug-nv , Thanks for the reviewing, I missed the global step will increase in batch_end when I adding this line, I have tested the use case of capturing single iteration, and confirmed you are right, fixed it

Comment (lilyw97) [reply nemo/core/classes/modelPT.py]: fixed it in the latest commit, resolve this conversation

File: nemo/core/classes/modelPT.py [hunk 7]
```diff
@@ -15,7 +15,9 @@
 
 import copy
 import inspect
+import json
```
[the remark this answers is not in the document]
Comment (titu1994) [reply nemo/core/classes/modelPT.py:18]: Could you remove unused imports ?

File: nemo/core/classes/modelPT.py [hunk 8]
```diff
@@ -1744,6 +1753,76 @@ def update_save_restore_connector(cls, save_restore_connector):
...
+
+                job_id = get_envint("SLURM_JOB_ID", 0)
+
+                self._chakra_trace_dir = trace_dir / f'{job_id}_chakra'
+                self._kineto_trace_dir = trace_dir / f'{job_id}_kineto'
+
+                self._chakra_trace_dir.mkdir(parents=True, exist_ok=True)
+                self._kineto_trace_dir.mkdir(parents=True, exist_ok=True)
+
+                if type(self._chakra_profile_start_step) == int:
```
Comment (titu1994) [inline nemo/core/classes/modelPT.py:1792]: Replace with is instance(...)

File: nemo/core/classes/modelPT.py [hunk 9]
```diff
@@ -1744,6 +1753,76 @@ def update_save_restore_connector(cls, save_restore_connector):
...
+                    logging.info(f'chakra profiling setup with start_step: {self._chakra_profile_start_step}')
+                else:
+                    raise ValueError(
+                        f'chakra start_step must be of type int. Found: {type(self._chakra_profile_start_step)}'
+                    )
+
```
Comment (titu1994) [inline nemo/core/classes/modelPT.py:1798]: ^

File: nemo/core/classes/modelPT.py [hunk 10]
```diff
@@ -1744,6 +1753,76 @@ def update_save_restore_connector(cls, save_restore_connector):
...
+                    pass
+                else:
+                    raise ValueError(f'chakra end_step must be greater than or equal to chakra start_step')
+
+                if self.cfg.nsys_profile.get('enabled', False):
+                    raise Exception(
+                        f"Profiler conflict: Chakra profiling and Nsys profiling cannot be enabled at the same time."
+                    )
+
+                self._et = ExecutionTraceObserver()
```
Comment (titu1994) [inline nemo/core/classes/modelPT.py:1816]: Q: is it necessary to import a private variable from torch profile? Move this optional import here and import only inside this block here so that if you don't use chakra and pytorch changes API, all Nemo models don't break due to model pt

Model is used by

File: nemo/core/classes/modelPT.py [hunk 11]
```diff
@@ -38,6 +40,7 @@
 from omegaconf import DictConfig, OmegaConf, open_dict
 from pytorch_lightning import LightningModule, Trainer
 from pytorch_lightning.utilities import model_summary, rank_zero_only
+from torch.profiler import ExecutionTraceObserver, _ExperimentalConfig
```
Comment (titu1994) [inline nemo/core/classes/modelPT.py:43]: Private import should be removed or at least moved directly to where it will be used. If PT changes API of private variable, it will break all Nemo models even if they don't use chakra or profiling

**Review by titu1994 — CHANGES_REQUESTED** [review]
Minor changes necessary, overall looks good

File: nemo/core/classes/modelPT.py (hunk 11 above)
Comment (lilyw97) [reply nemo/core/classes/modelPT.py]: thanks, removed to code block in **_setup_chakra_profiling** func in the latest commit

File: nemo/core/classes/modelPT.py (hunk 10 above)
Comment (lilyw97) [reply nemo/core/classes/modelPT.py:1815]: thanks for pointing it out, variable seems not necessary now, delete it , resolved in the latest commit

File: nemo/core/classes/modelPT.py (hunk 8 above)
Comment (lilyw97) [reply nemo/core/classes/modelPT.py]: thanks, resolved in the latest commit

**Review by maanug-nv — APPROVED** [review]

**Review by titu1994 — APPROVED** [review]
Looks good thanks !

**Comment by lilyw97** [comment]
Hi @maanug-nv , wondering how to trigger merging, I updated the branch, and seems there are still 2 checks that need to pass
